### ISSUE TITLE
TF apply after tainting an EC2 instance causing "resources changed outside of TF message"

### ISSUE THREAD

**Issue by Houlistonm** [hashicorp/terraform#29912]
### Terraform Version

```
Terraform v1.0.10
on linux_amd64
```

### Terraform Configuration Files

```terraform
No specific configuration
```

### Debug Output
None

### Expected Behavior
I was working on one of our ec2 instances debugging and issue and when I was done, I tainted the instances and ran TF apply.
After the apply finished, I immediately ran a plan expecting no changes (state == actual).

There are two generalized issues:
1. The `instance_id` of the ec2 instance changed as a result of the taint/apply, but the change was not reflected in the state for all of the resources associated with the `instance_id`
2. The `tags` in the root_block_device for the instance wasn't initialized (?)

### Actual Behavior

```
Terraform detected the following changes made outside of Terraform since the
last "terraform apply":

  # module.nat.module.eni.aws_network_interface.eni[2] has been changed
  ~ resource "aws_network_interface" "eni" {
        id                 = "eni-055f527de562db47e"
        tags               = {
            "foundation_name" = "ci"
        }
        # (9 unchanged attributes hidden)

      - attachment {
          - attachment_id = "eni-attach-040bd747ac1d70b8e" -> null
          - device_index  = 0 -> null
          - instance      = "i-068539baf025579a6" -> null
        }
      + attachment {
          + attachment_id = "eni-attach-091058123666aee45"
          + device_index  = 0
          + instance      = "i-0fa10106afc016944"
        }
    }
  # module.nat.module.eni.aws_eip.eip[2] has been changed
  ~ resource "aws_eip" "eip" {
        id                   = "eipalloc-07f0a846cd2d18241"
      ~ instance             = "i-068539baf025579a6" -> "i-0fa10106afc016944"
        tags                 = {
            "foundation_name" = "ci"
        }
        # (10 unchanged attributes hidden)

        # (1 unchanged block hidden)
    }
  # module.nat.module.eni.aws_eip_association.eip_association[2] has been changed
  ~ resource "aws_eip_association" "eip_association" {
        id                   = "eipassoc-0e8438738ba873a3b"
      ~ instance_id          = "i-068539baf025579a6" -> "i-0fa10106afc016944"
        # (4 unchanged attributes hidden)
    }

  # module.nat.module.nat_host.aws_instance.instance[2] has been changed
  ~ resource "aws_instance" "instance" {
        id                           = "i-0fa10106afc016944"
        tags                         = {
            "foundation_name"  = "ci"
            "job"              = "nat"
            "operating-system" = "Amazon Linux 2"
            "ssh_host_name"    = "ci_enterprise_services_nat_3"
        }
        # (27 unchanged attributes hidden)





      ~ root_block_device {
          + tags                  = {}
            # (9 unchanged attributes hidden)
        }
        # (4 unchanged blocks hidden)
    }
```

I would expect that TF and/or the AWS provider would account for these actions being caused by the taint/apply and not force a `tf apply -refresh-only` to clean up the status.  It's difficult to look at the output and quickly determine if there is a problem.

 Note: We repave our ec2 instances weekly.

### Steps to Reproduce
1. `cd top level of project`
2. `terragrunt taint 'module.nat.module.nat_host.aws_instance.instance[2]' --terragrunt-working-dir enterprise-services`
3. `terragrunt apply`
4. `terragrunt plan --terragrunt-working-dir enterprise-services`

and I was met with the following status:

**Comment by jbardin** [comment]
Hi @Houlistonm,

Thanks for filing the issue. Any changes reported as "changes made outside of Terraform" are strictly from the provider, and outside of the control of Terraform. I don't think tainting the resource is directly related, and normal replacement from a configuration change should have produced the same result.

It would be nice if we could confirm this however, before sending it back to the provider for triage. Can you supply the configuration used for the example, and possibly reduce it to a minimal reproduction without the use of terragrunt?

Thanks!

**Comment by Houlistonm** [comment]
## Sample config ##
```terraform
terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 3.35"
    }
  }
}

# Configure the AWS Provider
provider "aws" {
  region = "us-east-1"
}

resource "aws_vpc" "voldemort" {
  cidr_block = "10.0.0.0/16"
}

resource "aws_subnet" "subnet" {
  cidr_block = "10.0.1.0/24"
  vpc_id     = aws_vpc.voldemort.id

  availability_zone = "us-east-1a"
}

resource "aws_network_interface" "eni" {
  subnet_id = aws_subnet.subnet.id
}

resource "aws_instance" "test" {
  ami           = "ami-0331cc1e9ee0e02f5"
  instance_type = "t2.micro"
  network_interface {
    device_index         = 0
    network_interface_id = aws_network_interface.eni.id
  }
}
```

`terraform apply -auto-approve`

```
<REDACTED FOR BREVITY>
aws_vpc.voldemort: Creating...
aws_vpc.voldemort: Creation complete after 6s [id=vpc-0eba06b15d4ad95a7]
aws_subnet.subnet: Creating...
aws_subnet.subnet: Creation complete after 3s [id=subnet-0e5e3db9362638ae1]
aws_network_interface.eni: Creating...
aws_network_interface.eni: Still creating... [10s elapsed]
aws_network_interface.eni: Still creating... [20s elapsed]
aws_network_interface.eni: Still creating... [30s elapsed]
aws_network_interface.eni: Creation complete after 32s [id=eni-0cf7661ff6f315661]
aws_instance.test: Creating...
aws_instance.test: Still creating... [10s elapsed]
aws_instance.test: Still creating... [20s elapsed]
aws_instance.test: Still creating... [30s elapsed]
aws_instance.test: Still creating... [40s elapsed]
aws_instance.test: Still creating... [50s elapsed]
aws_instance.test: Still creating... [1m0s elapsed]
aws_instance.test: Still creating... [1m10s elapsed]
aws_instance.test: Creation complete after 1m19s [id=i-0a5f6593712089761]

Apply complete! Resources: 4 added, 0 changed, 0 destroyed.
```

`terraform plan`

```
aws_vpc.voldemort: Refreshing state... [id=vpc-0eba06b15d4ad95a7]
aws_subnet.subnet: Refreshing state... [id=subnet-0e5e3db9362638ae1]
aws_network_interface.eni: Refreshing state... [id=eni-0cf7661ff6f315661]
aws_instance.test: Refreshing state... [id=i-0a5f6593712089761]

Note: Objects have changed outside of Terraform

Terraform detected the following changes made outside of Terraform since the last "terraform apply":

  # aws_vpc.voldemort has been changed
  ~ resource "aws_vpc" "voldemort" {
        id                               = "vpc-0eba06b15d4ad95a7"
      + tags                             = {}
        # (15 unchanged attributes hidden)
    }
  # aws_instance.test has been changed
  ~ resource "aws_instance" "test" {
        id                           = "i-0a5f6593712089761"
      + tags                         = {}
        # (24 unchanged attributes hidden)





        # (5 unchanged blocks hidden)
    }
  # aws_network_interface.eni has been changed
  ~ resource "aws_network_interface" "eni" {
        id                 = "eni-0cf7661ff6f315661"
      + tags               = {}
        # (9 unchanged attributes hidden)

      + attachment {
          + attachment_id = "eni-attach-00f2a6ef0c25c9540"
          + device_index  = 0
          + instance      = "i-0a5f6593712089761"
        }
    }
  # aws_subnet.subnet has been changed
  ~ resource "aws_subnet" "subnet" {
        id                              = "subnet-0e5e3db9362638ae1"
      + tags                            = {}
        # (10 unchanged attributes hidden)
    }

Unless you have made equivalent changes to your configuration, or ignored the relevant attributes using ignore_changes, the following plan may include actions to undo or respond to these changes.

──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

No changes. Your infrastructure matches the configuration.

Your configuration already matches the changes detected above. If you'd like to update the Terraform state to match, create and apply a refresh-only plan:
  terraform apply -refresh-only
```

`terraform apply -refresh-only -auto-approve`

```
aws_vpc.voldemort: Refreshing state... [id=vpc-0eba06b15d4ad95a7]
aws_subnet.subnet: Refreshing state... [id=subnet-0e5e3db9362638ae1]
aws_network_interface.eni: Refreshing state... [id=eni-0cf7661ff6f315661]
aws_instance.test: Refreshing state... [id=i-0a5f6593712089761]

Note: Objects have changed outside of Terraform

Terraform detected the following changes made outside of Terraform since the last "terraform apply":

<REDACTED FOR BREVITY>

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
```

`terraform taint aws_instance.test`

`terraform apply -auto-approve`
```
aws_vpc.voldemort: Refreshing state... [id=vpc-0eba06b15d4ad95a7]
aws_subnet.subnet: Refreshing state... [id=subnet-0e5e3db9362638ae1]
aws_network_interface.eni: Refreshing state... [id=eni-0cf7661ff6f315661]
aws_instance.test: Refreshing state... [id=i-0a5f6593712089761]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
-/+ destroy and then create replacement

Terraform will perform the following actions:

  # aws_instance.test is tainted, so must be replaced
-/+ resource "aws_instance" "test" {
<REDACTED FOR BREVITY>
    }

Plan: 1 to add, 0 to change, 1 to destroy.
aws_instance.test: Destroying... [id=i-0a5f6593712089761]
aws_instance.test: Still destroying... [id=i-0a5f6593712089761, 10s elapsed]
aws_instance.test: Still destroying... [id=i-0a5f6593712089761, 20s elapsed]
aws_instance.test: Still destroying... [id=i-0a5f6593712089761, 30s elapsed]
aws_instance.test: Still destroying... [id=i-0a5f6593712089761, 40s elapsed]
aws_instance.test: Destruction complete after 42s
aws_instance.test: Creating...
aws_instance.test: Still creating... [10s elapsed]
aws_instance.test: Still creating... [20s elapsed]
aws_instance.test: Still creating... [30s elapsed]
aws_instance.test: Still creating... [40s elapsed]
aws_instance.test: Creation complete after 47s [id=i-0867e988959ea6025]

Apply complete! Resources: 1 added, 0 changed, 1 destroyed.
```
`terraform plan`
```
aws_vpc.voldemort: Refreshing state... [id=vpc-0eba06b15d4ad95a7]
aws_subnet.subnet: Refreshing state... [id=subnet-0e5e3db9362638ae1]
aws_network_interface.eni: Refreshing state... [id=eni-0cf7661ff6f315661]
aws_instance.test: Refreshing state... [id=i-0867e988959ea6025]

Note: Objects have changed outside of Terraform

Terraform detected the following changes made outside of Terraform since the last "terraform apply":

  # aws_network_interface.eni has been changed
  ~ resource "aws_network_interface" "eni" {
        id                 = "eni-0cf7661ff6f315661"
        tags               = {}
        # (9 unchanged attributes hidden)

      - attachment {
          - attachment_id = "eni-attach-00f2a6ef0c25c9540" -> null
          - device_index  = 0 -> null
          - instance      = "i-0a5f6593712089761" -> null
        }
      + attachment {
          + attachment_id = "eni-attach-0a384832f06334aba"
          + device_index  = 0
          + instance      = "i-0867e988959ea6025"
        }
    }
  # aws_instance.test has been changed
  ~ resource "aws_instance" "test" {
        id                           = "i-0867e988959ea6025"
      + tags                         = {}
        # (24 unchanged attributes hidden)





        # (5 unchanged blocks hidden)
    }

Unless you have made equivalent changes to your configuration, or ignored the relevant attributes using ignore_changes, the following plan may include actions to undo or respond to these changes.

──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

No changes. Your infrastructure matches the configuration.

Your configuration already matches the changes detected above. If you'd like to update the Terraform state to match, create and apply a refresh-only plan:
  terraform apply -refresh-only
```

**Comment by jbardin** [comment]
Thanks for the extra details @Houlistonm!

The plan outputs here more clearly show what is going wrong. The `aws_network_interface` attachments are computed by the provider, but because the provider is not indicating that they need to be changed during the initial plan, the values cannot be updated until the next plan cycle where they are refreshed to show the attached instances.

However, because the condition for this missing change is dependent on a separate resource, the provider doesn't really have a good way to predict when that change should be made. They likely attempt to set these values during apply but they are rejected because the changes do not match the planned set of values (this does not result in an error, due to more relaxed constraints when using the legacy SDK). 

There's not much we can do on this side within the current constraints of the provider protocol, but I'm going to mark this as an enhancement request for future thought. Because this is caused by an implied dependency not declared in the configuration, perhaps the answer here is to ensure there is a way to declare this in the configuration. This could be in the form of a lifecycle option similar to `depends_on`, or even at the provider level, where the `aws_network_interface` requires a list of instances that will be attached.

**Comment by apparentlymart** [comment]
I do agree with the analysis above, but just wanted to note that this does seem to be a particularly quirky case because the provider seems to be modelling that relationship in both directions: the `aws_instance` resource refers to the network interface _and_ the `aws_network_interface` refers to the instance, and so the only way for this to resolve itself would be to revisit `aws_network_interface.eni` and refresh it _again_ after the `aws_instance.test` resource's changes are applied.

Today's Terraform typically visits each resource instance only once to refresh/plan (during planning) and once to apply (during applying), so there's nowhere in the lifecycle where we could resolve a cyclic dependency like this. We could potentially re-refresh all objects after their dependents have had changes applied, but that would be a very expensive answer to a pretty rare situation. :confounded: 

I think it could be helpful to try to understand here why/whether this unusual circular relationship is a useful thing to express inside Terraform. Typically we expect relationships to point in only one direction, which therefore avoids this problem of needing to revisit objects we already dealt with earlier in order to react to changes moving "backwards" through the dependency graph. If we can understand why such a circular dependency is useful to model in Terraform then we might be able to find other examples of it to use as the basis for designing a more general solution to it. :thinking:

**Comment by Houlistonm** [comment]
I like the idea of the post-apply "re-rerefresh".  I understand there is a cost associated with it and I'd be happy if this were a command line switch that we could enable/opt-in.  

Honestly, we're becoming numb to all of the "changed outside of terraform" messages.  The repaving of our instances touches many different resources (network_attachment, eip/eip_association, storage, route tables, load balancers, null resource ([separate issue](https://github.com/hashicorp/terraform/issues/24675#issue-600195510)) ), and that's just from memory.

**Comment by apparentlymart** [comment]
If your configuration's refresh step is cheap enough that this wouldn't add a burdensome amount of extra time to a run, you could perhaps today run `terraform apply -refresh-only -auto-approve` immediately after each `terraform apply` to get the effect of a naive "just refresh everything to make sure" approach, though I acknowledge that is only a helpful answer if you were already running Terraform in automation anyway and so can customize that automation to do that extra step.

I think for that to be a viable general solution we'd need to have some way to detect (or explicitly opt-in to, though we always consider that to be a last resort) the exceptional cases where it's necessary rather than just naively refreshing _everything_ again, because refreshing can be a pretty significant part of the time taken to produce a plan in typical configurations.

**Comment by jbardin** [comment]
Thanks @apparentlymart, I overlooked the reference in the config so didn't see the circular dependency this would cause. In that case there's really not much we could do via configuration, because any configuration which could describe this dependency would be invalid.

It would also be good to note that if in the future were `aws_network_interface` migrated off the legacy SDK as-is, attempting to include the attachments like this would fail when the applied state doesn't match the plan. The relationship would have to be re-thought in some way, or perhaps the attachments serve no real purpose in the managed resource and could be skipped altogether.
